src/ctapipe/calib/camera/extractor.py

```diff
@@ -0,0 +1,426 @@
+
+        Returns
+        -------
+        arrival_time_stats : dict
+        The sample statistics of the arrival time of each camera pixel.
+        """
+
+        # ensure numpy array
+        masked_arrival_times = np.ma.array(arrival_times, mask=broken_pixels)
+
+        # median over the sample per pixel
+        pixel_median = np.ma.median(masked_arrival_times, axis=0)
+
+        # mean over the sample per pixel
+        pixel_mean = np.ma.mean(masked_arrival_times, axis=0)
+
+        # std over the sample per pixel
+        pixel_std = np.ma.std(masked_arrival_times, axis=0)
+
+        # median of the median over the camera
+        median_of_pixel_median = np.ma.median(pixel_median, axis=1)
+
+        # time outliers from median
+        relative_median = pixel_median - median_of_pixel_median[:, np.newaxis]
+        time_median_outliers = np.logical_or(
+            pixel_median < self.time_cut_outliers[0],
+            pixel_median > self.time_cut_outliers[1],
+        )
+
+        return {
+            "time_mean": pixel_mean.filled(np.nan) * u.ns,
+            "time_median": pixel_median.filled(np.nan) * u.ns,
+            "time_std": pixel_std.filled(np.nan) * u.ns,
+            "relative_time_median": relative_median.filled(np.nan) * u.ns,
+            "time_median_outliers": time_median_outliers.filled(True),
+        }
+
+    def _clean_buffer(self):
+        """
+        Clean up the buffer by removing the events that will not participate
+        in the caluclation of the next sample.
+        """
+        self.n_events_in_buffer -= self.update_frequency
+        self.trigger_time = self.trigger_time[self.update_frequency :]
+        self.charges = self.charges[self.update_frequency :]
+        self.arrival_times = self.arrival_times[self.update_frequency :]
+        self.broken_pixels = self.broken_pixels[self.update_frequency :]
+
+    @abstractmethod
+    def _process(self, event: ArrayEventContainer):
+        """
+        Process the sample statistics of calibration events and fill the
+        corresponding MonitoringCameraContainer.
+
+        Parameters
+        ----------
+        event : ArrayEventContainer
+        """
+
+
+class FlatFieldExtractor(CalibrationExtractor):
+    """
+    Extractor for flat-field events.
+    """
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+    def _process(self, event: ArrayEventContainer):
+        """
+        Process the sample statistics of flat-field events and fill the
+        mon.tel[self.tel_id].flatfield container
+        """
+
+        container = event.mon.tel[self.tel_id].flatfield
+
+        charge_stats = self._calculate_charge_stats(
+            self.charges,
+            self.sample_masked_pixels,
+            self.trigger_times,
+        )
+        arrivaltime_stats = self._calculate_arrival_time_stats(
+            self.arrival_times,
+            self.sample_masked_pixels,
+        )
+
+        flatfield_container = {
```

Comment:
We made very bad experiences with modifying containers in place. You should create and fill a fresh container.

You can also directly pass the fields to the container itself.